src/Raven.NewClient/Document/AsyncHiLoKeyGenerator.cs

```diff
@@ -23,193 +17,124 @@ namespace Raven.NewClient.Client.Document
     /// </summary>
     public class AsyncHiLoKeyGenerator : HiLoKeyGeneratorBase
     {
-        private SpinLock generatorLock = new SpinLock(enableThreadOwnerTracking: false); // Using a spin lock rather than Monitor.Enter, because it's not reentrant
-
         /// <summary>
         /// Initializes a new instance of the <see cref="HiLoKeyGenerator"/> class.
         /// </summary>
-        public AsyncHiLoKeyGenerator(string tag, long capacity)
-            : base(tag, capacity)
+        public AsyncHiLoKeyGenerator(string tag, DocumentStore store, string dbName, string identityPartsSeparator)
+            : base(tag, store, dbName, identityPartsSeparator)
         {
         }
 
         /// <summary>
         /// Generates the document key.
         /// </summary>
-        /// <param name="databaseCommands">The commands.</param>
-        /// <param name="convention">The convention.</param>
         /// <param name="entity">The entity.</param>
         /// <returns></returns>
-        public Task<string> GenerateDocumentKeyAsync(DocumentConvention convention, object entity)
+        public Task<string> GenerateDocumentKeyAsync(object entity)
         {
-            return NextIdAsync().ContinueWith(task => GetDocumentKeyFromId(convention, task.Result));
+            return NextIdAsync().ContinueWith(task => GetDocumentKeyFromId(task.Result));
         }
 
         ///<summary>
         /// Create the next id (numeric)
         ///</summary>
         public Task<long> NextIdAsync()
         {
-            var myRange = Range; // thread safe copy
-            long incrementedCurrent = Interlocked.Increment(ref myRange.Current);
-            if (incrementedCurrent <= myRange.Max)
+            while (true)
             {
-                return CompletedTask.With(incrementedCurrent);
-            }
+                var myRange = Range; // thread safe copy
+                var incrementedCurrent = Interlocked.Increment(ref myRange.Current);
+                if (incrementedCurrent <= myRange.Max)
+                {
+                    return CompletedTask.With(incrementedCurrent);
+                }
 
-            bool lockTaken = false;
-            try
-            {
-                generatorLock.Enter(ref lockTaken);
-                if (Range != myRange)
+                if (interlockedLock.TryEnter() == false)
                 {
-                    // Lock was contended, and the max has already been changed. Just get a new id as usual.
-                    generatorLock.Exit();
-                    return NextIdAsync();
+                    Interlocked.Increment(ref threadsWaitingForRangeUpdate);
+                    mre.WaitOne();
```

Comment:
You are in the middle of an async task, and you are doing a sync wait here, that is a bad idea.